### ISSUE TITLE
Can `static T` be used as a generic proc parameter?

### ISSUE THREAD

**Issue by awr1** [nim-lang/Nim#14941]
[The manual does not currently clarify this](https://nim-lang.org/docs/manual.html#special-types), so this might be a bug.

```nim
type Foo[T :static[int]] = object
proc bar[T :static[int]](x :T) :Foo[T] = discard
```
The signature of `bar` breaks, however:
```nim
proc baz(x :static[int]) :Foo[x] = discard
proc bah[F :Foo](x :F.T) :F = discard
```
both validate.

**Comment by Clyybber** [comment]
I'd say this is expected.
```nim
type Foo[T :static[int]] = object
proc bar[T :static[int]](x :type T) :Foo[T] = discard
```
works.

I think the reason is that `[T: static int]` doesn't mean that `T` *is* the type `static int`, but means that `T` is of the type `static int` (I'm not entirely sure that this is the correct way to think about it, since `type static int` is `static int`).

**Comment by zah** [comment]
I agree that there is a bug here. The following program should not compile:

```nim
type Foo[N :static[int]] = object

proc baz(n: static[int]): Foo[n] = discard
proc bah[F: Foo](x: F.N): F = discard # There should be an error here
```

Accessing the static generic param through the dot operator should work in the same way in any context and it should return a value, not a type:

```nim
type F5 = Foo[5]
echo F5.N # This should print 5
```

Let's now imagine `bah` being instantiated with a concrete `Foo` type (e.g. `Foo[5]`). Its signature becomes:

```nim
proc bah(x: 5): Foo[5]
```

... which should result in the error "type expected, but got: 5". 

Araq has sometimes pushed for a relaxation in the type system allowing the compiler to interpret `5` as a type meaning "allow only the value 5 to be passed here", but that requires a RFC.process.
